romancal/source_detection/source_detection_step.py

```diff
@@ -140,17 +160,54 @@ def process(self, input):
                     dtype=(int, np.float64, np.float64, np.float64),
                 )
 
-            # attach source catalog to output model as array in meta.source_detecion
-            # the table will be stored as a 1D array with the four columns
-            # concatenated, in order, with units attached
-            catalog_as_array = np.array(
-                [
-                    catalog["id"].value,
-                    catalog["xcentroid"].value,
-                    catalog["ycentroid"].value,
-                    catalog["flux"].value,
-                ]
-            )
+            if self.fit_psf and len(sources):
+                # refine astrometry by fitting PSF models to the detected sources
+                log.info("Constructing a gridded PSF model.")
+                detector = input_model.meta.instrument["detector"].replace("WFI", "SCA")
+                gridded_psf_model, _ = psf.create_gridded_psf_model(
+                    path_prefix="tmp",
+                    filt=filt,
+                    detector=detector,
+                    overwrite=True,
+                    logging_level="ERROR",
+                )
+
+                log.info(
+                    "Fitting a PSF model to sources for improved astrometric precision."
+                )
+                psf_photometry_table, photometry = psf.fit_psf_to_image_model(
+                    image_model=input_model,
+                    psf_model=gridded_psf_model,
+                    x_init=catalog["xcentroid"],
+                    y_init=catalog["ycentroid"],
+                    exclude_out_of_bounds=True,
+                )
+
+                # attach source catalog to output model as array in
+                # meta.source_detection the table will be stored as a
+                # 1D array with the four columns concatenated, in order,
+                # with units attached
+                catalog_as_array = np.array(
+                    [
+                        psf_photometry_table["id"].value,
+                        psf_photometry_table["x_fit"].value,
+                        psf_photometry_table["y_fit"].value,
+                        psf_photometry_table["flux_fit"].value,
+                    ]
+                )
+            else:
+                # attach source catalog to output model as array in
+                # meta.source_detection the table will be stored as a
+                # 1D array with the four columns concatenated, in order,
+                # with units attached
+                catalog_as_array = np.array(
+                    [
+                        catalog["id"].value,
+                        catalog["xcentroid"].value,
+                        catalog["ycentroid"].value,
+                        catalog["flux"].value,
+                    ]
```

Comment:
I don't know how hard it is to check, and maybe we defer this to another PR, but we really want to use at least a structured array here eventually, as well as saving all the other columns that may have value.

Reply:
I will ping @PaulHuwe on Slack to ask about this.

Reply:
I confirmed that structured arrays are supported in ASDF. You can test for yourself with:
```python
import numpy as np
import roman_datamodels.datamodels as rdd
from roman_datamodels.maker_utils import mk_level2_image

im = rdd.ImageModel(mk_level2_image())

idx = np.random.randint(0, 100, 10)
x = np.random.uniform(0, 100, 10)
y = np.random.uniform(0, 100, 10)

recarr = np.core.records.fromarrays(
    [idx, x, y], 
    names="idx, x, y", 
    formats=[int, float, float]
)

im.meta['test_recarr'] = recarr

im.to_asdf('test_recarr.asdf')

im_reloaded = rdd.open('test_recarr.asdf')

recarr2 = im_reloaded.meta['test_recarr']

assert np.all(recarr2 == recarr)
```

I propose that we make a follow-up PR for the move to structured arrays, since that will touch tweakreg a bunch. Thoughts?

Reply:
That sounds good, thanks.

Reply:
As long as no one cares about accessing this information as text in the metadata, then this is great.

Reply:
By default, this catalog is later deleted by tweakreg, so it shouldn't be useful to anyone unless they stop the pipeline part-way through.

Reply:
We definitely don't want a potentially huge catalog converted to text anywhere!